scipy/integrate/_rules/_base.py

```diff
@@ -447,50 +456,72 @@ def _subregion_coordinates(a, b):
         ([1/2, 1/2], [1, 1])
     """
 
-    m = (a + b)/2
+    xp = array_namespace(a, b)
+    m = (a + b) * 0.5
+
+    left = [xp.asarray([a[i], m[i]]) for i in range(a.shape[0])]
+    right = [xp.asarray([m[i], b[i]]) for i in range(b.shape[0])]
+
+    a_sub = _cartesian_product(left)
+    b_sub = _cartesian_product(right)
 
-    for a_sub, b_sub in zip(
-        itertools.product(*np.array([a, m]).T),
-        itertools.product(*np.array([m, b]).T)
-    ):
-        yield np.array(a_sub), np.array(b_sub)
+    for i in range(a_sub.shape[0]):
+        yield a_sub[i, ...], b_sub[i, ...]
 
 
 def _apply_fixed_rule(f, a, b, orig_nodes, orig_weights, args=()):
+    xp = array_namespace(a, b, orig_nodes, orig_weights)
+
     # Ensure orig_nodes are at least 2D, since 1D cubature methods can return arrays of
     # shape (npoints,) rather than (npoints, 1)
     if orig_nodes.ndim == 1:
-        orig_nodes = orig_nodes[:, np.newaxis]
+        orig_nodes = orig_nodes[:, None]
 
     rule_ndim = orig_nodes.shape[-1]
 
-    a_ndim = len(a)
-    b_ndim = len(b)
+    a_ndim = xp_size(a)
+    b_ndim = xp_size(b)
 
     if rule_ndim != a_ndim or rule_ndim != b_ndim:
         raise ValueError(f"rule and function are of incompatible dimension, nodes have"
                          f"ndim {rule_ndim}, while limit of integration has ndim"
                          f"a_ndim={a_ndim}, b_ndim={b_ndim}")
 
+    a = xp.astype(a, xp.float64)
+    b = xp.astype(b, xp.float64)
     lengths = b - a
 
     # The underlying rule is for the hypercube [-1, 1]^n.
     #
     # To handle arbitrary regions of integration, it's necessary to apply a linear
     # change of coordinates to map each interval [a[i], b[i]] to [-1, 1].
-    nodes = (orig_nodes + 1) * (lengths / 2) + a
+    nodes = (orig_nodes + 1) * (lengths * 0.5) + a
 
     # Also need to multiply the weights by a scale factor equal to the determinant
     # of the Jacobian for this coordinate change.
     weight_scale_factor = math.prod(lengths) / 2**rule_ndim
     weights = orig_weights * weight_scale_factor
 
     f_nodes = f(nodes, *args)
-    weights_reshaped = weights.reshape(-1, *([1] * (f_nodes.ndim - 1)))
+    weights_reshaped = xp.reshape(weights, (-1, *([1] * (f_nodes.ndim - 1))))
 
     # f(nodes) will have shape (num_nodes, output_dim_1, ..., output_dim_n)
     # Summing along the first axis means estimate will shape (output_dim_1, ...,
     # output_dim_n)
-    est = np.sum(weights_reshaped * f_nodes, axis=0)
+    est = xp.sum(weights_reshaped * f_nodes, axis=0)
 
     return est
+
+
+def _concat(arrays, axis=0):
+    xp = array_namespace(*arrays)
+
+    if hasattr(xp, "concat"):
+        return xp.concat(arrays, axis=axis)
+    elif hasattr(xp, "concatenate"):
+        return xp.concatenate(arrays, axis=axis)
+    else:
+        return xp.asarray(np_compat.concatenate(
+            [np_compat.asarray(arr) for arr in arrays],
+            axis=axis,
+        ))
```

Comment:
Thanks for taking a look, I've removed this now. The problem was that when each a `Rule` like `GaussKronrodQuadrature` is initialised, there's no array passed in the constructor that can be used to determine `xp`. Instead each `Rule` has an extra `xp` parameter in its `__init__`:

```python
def __init__(self, npoints, xp=None):
    # ...
    if xp is None:
        xp = np_compat

    self.xp = xp
```

Then it can return nodes and weights in the correct array namespace. But when the tests set NumPy as the backend, it was initialising each rule with `np` rather than `np_compat`. I've modified the constructors for each of the rules so that it will check if `xp` is NumPy and set it to `np_compat` instead. Hopefully this is the right approach?

Reply:
Sounds like a good approach yes, although why do we need to add `or is_numpy(xp)` to the condition? Either `xp=None`, or one should pass in `np_compat === array_namespace(an_np_array_or_scalar)`.

Reply:
If the rule is initialised like `Rule(..., xp=np)` for NumPy version < 2 (as it is in tests by the `@array_api_compatible` decorator) then `xp.concat` won't exist.

Reply:
Then you may also need similar shims for other libraries. I think you'll either want to:

- require that a compatible namespace is passed to `xp`, and do so in the tests
- acquire a compatible namespace in the rules via `xp = array_namespace(xp.empty(0))`

Reply:
Thanks! I'll go for option number 2, I was just about to update my comment asking if there was a way of determining the compatible namespace like this.